### PR TITLE
Don't use conditionals inside the head

### PR DESCRIPTION
**Pull request by arcataroger**
It looks like we're experiencing a hydration error with the Head component in the Svelte Starter Kit (https://www.datocms.com/marketplace/starters/sveltekit-starter-kit) when loading. 

It could be that the conditionals are causing Svelte to generate some HTML comments as hydration markers, which might result in a mismatch between what the server outputs and what the client sees.

### CODE REVIEW

**Comment by emileswain** [comment]
Encountered a related issue perhaps when upgrading to svelte5 and using the Head component. THe meta tags are not closing correctly. 

It might be related to this nuance https://svelte.dev/docs/svelte/v5-migration-guide#Other-breaking-changes-null-and-undefined-become-the-empty-string where null values are perhaps being converted to empty strings?

**Comment by arcataroger** [comment]
Thank you @emileswain! We will investigate.

@stefanoverna Left you details in Basecamp

File: src/lib/components/Head/Head.svelte [hunk 1]
```diff
@@ -7,63 +7,64 @@
 		/** the HTML attributes to attach to the meta tag */
 		attributes?: Record<string, string> | null | undefined;
 	}
-
```
Comment (stefanoverna) [inline src/lib/components/Head/Head.svelte:10]: can you please keep the newlines? thanks!

File: src/lib/components/Head/Head.svelte [hunk 2]
```diff
@@ -7,63 +7,64 @@
...
 </script>
 
 <script lang="ts">
 	export let data: Array<TitleMetaLinkTag | SeoOrFaviconTag> = [];
+	// Statically output HTML instead of using conditionals inside <svelte:head>
+	const renderedTags = data
+		.map(({ tag, attributes, content }) => {
+			const attrs = attributes
+				? Object.entries(attributes)
+						.map(([key, value]) => `${key}="${value}"`)
```
Comment (stefanoverna) [inline src/lib/components/Head/Head.svelte:56]: this is brittle.. I'd feel more safe with a utility like this one that escapes HTML, PLUS some tests that verify that the encoding is working as expected:

```ts
export const tag = (
  tagName: string,
  attrs: Record<string, string | null | undefined>,
  content?: Array<string | null | undefined> | null,
) => {
  const serializedAttrs = [];

  if (attrs) {
    for (const attrName in attrs) {
      if (Object.prototype.hasOwnProperty.call(attrs, attrName)) {
        const value = attrs[attrName];
        if (value) {
          serializedAttrs.push(
            `${escapeString(attrName)}="${escapeString(value)}"`,
          );
        }
      }
    }
  }

  const attrsString =
    serializedAttrs.length > 0 ? ` ${serializedAttrs.join(' ')}` : '';

  return content
    ? `<${tagName}${attrsString}>${content.join('')}</${tagName}>`
    : `<${tagName}${attrsString} />`;
};

// Example usage:

tag('source', {
  srcset: this.data.webpSrcSet,
  sizes: this.sizes ?? this.data.sizes ?? undefined,
  type: 'image/webp',
});
```

**Review by stefanoverna — CHANGES_REQUESTED** [review]

File: src/lib/components/Head/Head.svelte (hunk 1 above)
Comment (arcataroger) [reply src/lib/components/Head/Head.svelte:10]: Oh sorry! Weird, I thought that's something Prettier automatically does. TIL!

File: src/lib/components/Head/Head.svelte (hunk 2 above)
Comment (arcataroger) [reply src/lib/components/Head/Head.svelte:56]: @stefanoverna good changes, ty! ^

But are we sure this is the even the correct approach to this issue? I'm not familiar enough with Svelte to know if this even the right way to fix this bug. The documentation was really lacking and I found a lot of conflicting info online... nothing else I tried worked, though, except for this.

But it seems strange that you'd need to make a string literal just to avoid a hydration error. I thought Svelte should be pretty mature by now, but if we can't use conditionals in the head... shouldn't a linter rule or such detect that...?

Comment (stefanoverna) [reply src/lib/components/Head/Head.svelte:56]: Conditionals should work, but they don't and we're not doing anything strange here. We can proceed in parallel with an issue on their end, and in the meantime provide this as an hot fix to avoid problems for our people :)

Comment (emileswain) [reply src/lib/components/Head/Head.svelte:56]: I took a look by copying the Head component. There is nothing with the conditional itself. i.e. content being null, undefined, '' etc, works as expected in svelte as a test and also your component is also doing what you would expect. 

The issue is i believe something to do with how <svelte:element> is behaving and there is some sort of weird behaviour that occurs when it trys to work out if it should close the tag or not. 

Try this, if you have both conditions running, then the meta tags don't close correctly. I.e. the No content condition renders are <meta></meta> 

If however you comment out the with content line,  then the No content will close correctly. 

Its like, svelte element isn't treating each tag independently, if even one element doesn't close, i.e. say you have a title tag. then none of the close correctly.

I evevn tried using snippets and seperate components. So this may be an issue with svelte/element and perhaps the meta tag and client side hydration perhaps? Its beyond me at this point.

File: src/lib/util/headTagToEscapedStrings.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+export const serializeHeadTag = (metaTag: SupportedHeadTags[number]) => {
+	const { tag, attributes, content } = metaTag;
+
+	const serializedAttributes: string[] =
+		attributes && typeof attributes === 'object'
+			? Object.entries(attributes).flatMap(([key, value]) =>
+					value ? `${escapeHtmlString(key)}="${escapeHtmlString(value)}"` : []
+			  )
+			: [];
```
Comment (arcataroger) [inline src/lib/util/headTagToEscapedStrings.ts:21]: I think this is more modern & easier to read than:

```js
const serializedAttrs = [];

if (attrs) {
  for (const attrName in attrs) {
    if (Object.prototype.hasOwnProperty.call(attrs, attrName)) {
      const value = attrs[attrName];
      if (value) {
        serializedAttrs.push(
          `${escapeString(attrName)}="${escapeString(value)}"`,
        );
      }
    }
  }
}
```

Is that OK? Object.entries should already only enumerate only an object's own props, not anything it inherits: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries#description

It should supercede both `Object.hasOwnProperty.call()` and `Object.hasOwn()` (es2022)

File: src/lib/util/escapeHtmlString.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+/** Replaces special chars (&<>"') with HTML entities */
+export const escapeHtmlString = (html: string) => {
+	const escaped = html
+		.replace(
+			/&(?!amp;|lt;|gt;|quot;|apos;)/g, // Don't re-encode these entities
```
Comment (arcataroger) [inline src/lib/util/escapeHtmlString.ts:5]: edge case caught by a test; hopefully unlikely in real-world usage

File: src/lib/util/escapeHtmlString.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+/** Replaces special chars (&<>"') with HTML entities */
+export const escapeHtmlString = (html: string) => {
```
Comment (arcataroger) [inline src/lib/util/escapeHtmlString.ts:2]: I think we have several copies of this func (or very similar ones) across several repos? Not a high priority since it's so simple, but maybe we should extract it out into a `@datocms/utils` lib. Maybe also test it against a proper HTML sanitization lib to make sure we're covering all the cases?

File: src/lib/components/Head/Head.svelte (hunk 2 above)
Comment (arcataroger) [reply src/lib/components/Head/Head.svelte]: @stefanoverna Added some tests and cleaned it up a bit. Big change at https://github.com/datocms/datocms-svelte/pull/14/files?diff=split&w=1#r1876437362 and minor ones throughout.

@emileswain, I think we should open a new issue with Svelte directly with a minimal example and discuss the `<svelte:element>` stuff there. I'll open an issue and tag you there. TY again for the help here!

**Comment by arcataroger** [comment]
@stefanoverna Added some tests. One big change at https://github.com/datocms/datocms-svelte/pull/14/files?diff=split&w=1#r1876437362 and other minor cleanup throughout.

@emileswain For now, in here, we'll just patch around it by serializing the head. But I think we should open a new issue with Svelte directly to continue discussing the `<svelte:element>` issues with a minimal example. I'll do that in a bit and tag you there. TY again for the help here!

**Comment by emileswain** [comment]
> @stefanoverna Added some tests. One big change at https://github.com/datocms/datocms-svelte/pull/14/files?diff=split&w=1#r1876437362 and other minor cleanup throughout.
> 
> @emileswain For now, in here, we'll just patch around it by serializing the head. But I think we should open a new issue with Svelte directly to continue discussing the `<svelte:element>` issues with a minimal example. I'll do that in a bit and tag you there. TY again for the help here!

Makes sense, fyi, sorting the seotags so content elements come last, worked as a temporary fix.

**Review by stefanoverna — APPROVED** [review]

**Comment by stefanoverna** [comment]
Thanks, merged and squashed as https://github.com/datocms/datocms-svelte/commit/ec728a18e9b43b3dc5e2db17861c21d70e119ccb

**Comment by arcataroger** [comment]
@emileswain FYI I was trying to replicate this in a fresh minimal example for the Svelte team, but I could not.

Turns out our starter kit just had some old dependencies. Upgrading `vite` and `vite-plugin-svelte` fixes the hydration error, both with or without my hacky workaround above. Just FYI. Sorry about that.

If you already implemented  `"@datocms/svelte": "^4.0.1",`, that should still keep working.

**Comment by emileswain** [comment]
Hmm interesting,I did upgrade those packages as part of my upgrade to
svelt5. Which is when I encountered the issue.

I’ll check my project and see if I still have the problem.

Thanks
E

**Comment by arcataroger** [comment]
> Hmm interesting,I did upgrade those packages as part of my upgrade to svelt5. Which is when I encountered the issue.

Dumb question, but did you `npm install` after upgrading them, and were there any dependency issues? On our starter kit, it wouldn't let me seamlessly upgrade in-place because some dependency conflicts. I had to manually clear node_modules and regenerate the lockfile before the upgrade would go through. I think VScode doesn't always make it clear when the package.json doesn't match the actual installed versions?

(Sorry if that was obvious! Just making sure)

**Comment by emileswain** [comment]
I checked again and noticed that I may have had the component embedded
twice. Once in the page and a second time in the layout. So it might have
been user error on my part. Haven’t had a chance to double check, but might
explain it.

**Comment by arcataroger** [comment]
No worries. Hopefully the new version and upgrades will still fix it, either way. But if you do run into issues again, please let us know!
